pkg/ec2pricing/odpricing.go

```diff
@@ -0,0 +1,265 @@
+// Copyright Amazon.com Inc. or its affiliates. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"). You may
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package ec2pricing
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"path/filepath"
+	"strconv"
+	"time"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/endpoints"
+	"github.com/aws/aws-sdk-go/service/pricing"
+	"github.com/aws/aws-sdk-go/service/pricing/pricingiface"
+	"github.com/mitchellh/go-homedir"
+	"github.com/patrickmn/go-cache"
+	"go.uber.org/multierr"
+)
+
+const (
+	ODCacheFileName = "on-demand-pricing-cache.json"
+)
+
+type OnDemandPricing struct {
+	Region         string
+	FullRefreshTTL time.Duration
+	DirectoryPath  string
+	cache          *cache.Cache
+	pricingClient  pricingiface.PricingAPI
+}
+
+func LoadODCacheOrNew(pricingClient pricingiface.PricingAPI, region string, fullRefreshTTL time.Duration, directoryPath string) *OnDemandPricing {
+	expandedDirPath, err := homedir.Expand(directoryPath)
+	if err != nil {
+		log.Printf("Unable to load on-demand pricing cache directory %s: %v", expandedDirPath, err)
+		return &OnDemandPricing{
+			Region:         region,
+			FullRefreshTTL: 0,
+			DirectoryPath:  directoryPath,
+			cache:          cache.New(fullRefreshTTL, fullRefreshTTL),
+			pricingClient:  pricingClient,
+		}
+	}
+	odPricing := &OnDemandPricing{
+		Region:         region,
+		FullRefreshTTL: fullRefreshTTL,
+		DirectoryPath:  expandedDirPath,
+		pricingClient:  pricingClient,
+		cache:          cache.New(fullRefreshTTL, fullRefreshTTL),
+	}
+	if fullRefreshTTL <= 0 {
+		odPricing.Clear()
+		return odPricing
+	}
+	// Start the cache refresh job
```

Comment:
could you add why a goroutine vs. regular invocation in the comment? unless i'm missing something obvious..

Reply:
This is for a long running invocation of instance-selector, like a TUI, which might be coming soon...